src/main/java/com/playmonumenta/networkrelay/RemotePlayerManager.java

```diff
@@ -313,21 +329,33 @@ private static void remotePlayerChange(JsonObject data) {
 			if (shardRemotePlayers != null) {
 				shardRemotePlayers.put(remotePlayer.mName, remotePlayer);
 			}
-			CustomLogger.getInstance().ifPresent(logger -> logger.fine("Registering local player from remote data"));
+			MMLog.fine("Registering remote player " + remotePlayer.mName);
 			mRemotePlayersByUuid.put(remotePlayer.mUuid, remotePlayer);
 			mRemotePlayersByName.put(remotePlayer.mName, remotePlayer);
+			if (!remotePlayer.mIsHidden) {
+				mVisiblePlayers.add(remotePlayer.mUuid);
+			}
 			RemotePlayerLoadedEvent remotePLE = new RemotePlayerLoadedEvent(remotePlayer);
 			Bukkit.getServer().getPluginManager().callEvent(remotePLE);
+			return;
+		}
+
+		@Nullable Player localPlayer = Bukkit.getPlayer(remotePlayer.mUuid);
+		if (localPlayer != null && localPlayer.isOnline()) {
+			// Player logged off on remote shard, but is locally online.
+			// This can happen if the remote shard was not notified the player logged in here in time.
+			MMLog.fine("Detected race condition, triggering refresh on " + remotePlayer.mName);
+			refreshLocalPlayer(localPlayer);
```

Comment:
I did see the special case for when a local player quits, but is online on another shard. This is a fail safe for if the quit event arrives before that shard knows the player logged in elsewhere, which is what made this a race condition.